opm/autodiff/MultisegmentWells.hpp

```diff
@@ -0,0 +1,246 @@
+#include <opm/autodiff/LinearisedBlackoilResidual.hpp>
+#include <opm/autodiff/WellHelpers.hpp>
+
+#include <opm/autodiff/WellMultiSegment.hpp>
+
+
+
+namespace Opm {
+
+
+    /// Class for handling the multi-segment well model
+    class MultisegmentWells {
+    public:
+            // ---------      Types      ---------
+            using ADB = AutoDiffBlock<double>;
+            using Vector = ADB::V;
+
+            // Well operations and data needed.
+            struct MultisegmentWellOps {
+                explicit MultisegmentWellOps(const std::vector<WellMultiSegmentConstPtr>& wells_ms);
+                Eigen::SparseMatrix<double> w2p;              // well -> perf (scatter)
+                Eigen::SparseMatrix<double> p2w;              // perf -> well (gather)
+                Eigen::SparseMatrix<double> w2s;              // well -> segment (scatter)
+                Eigen::SparseMatrix<double> s2w;              // segment -> well (gather)
+                Eigen::SparseMatrix<double> s2p;              // segment -> perf (scatter)
+                Eigen::SparseMatrix<double> p2s;              // perf -> segment (gather)
+                Eigen::SparseMatrix<double> s2s_inlets;       // segment -> its inlet segments
+                Eigen::SparseMatrix<double> s2s_outlet;       // segment -> its outlet segment
+                Eigen::SparseMatrix<double> topseg2w;         // top segment -> well
+                AutoDiffMatrix eliminate_topseg;              // change the top segment related to be zero
+                std::vector<int> well_cells;                  // the set of perforated cells
+                Vector conn_trans_factors;                         // connection transmissibility factors
+                bool has_multisegment_wells;                  // flag indicating whether there is any muli-segment well
+            };
+
+            // copied from BlackoilModelBase
+            // should put to somewhere better
+            using DataBlock =  Eigen::Array<double,
+                                            Eigen::Dynamic,
+                                            Eigen::Dynamic,
+                                            Eigen::RowMajor>;
+
+            // ---------  Public methods  ---------
+            // TODO: using a vector of WellMultiSegmentConstPtr for now
+            // TODO: it should use const Wells or something else later.
+            MultisegmentWells(const std::vector<WellMultiSegmentConstPtr>& wells_multisegment, const int np);
+
+            const std::vector<WellMultiSegmentConstPtr>& wells() const;
+            const MultisegmentWellOps& wellOps() const;
+
+            int numPhases() const { return num_phases_; };
+
+            int numSegment() const { return nseg_total_; };
+
+            int numPerf() const { return nperf_total_; };
+
+            const Vector& wellPerforationCellPressureDiffs() const { return well_perforation_cell_pressure_diffs_; };
```

Comment:
Most of them are not, and even should be `protected`.  Some are just needed when they are still in the model class. I will fix them.